infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -708,4 +742,46 @@ public void shutdown() {
     private String getLocalHost() {
         return NodeLocator.parseString(serverContext.getLocalEndpoint()).getHost();
     }
+
+    private LockClient createLockClient(LockStateType initState) throws Exception {
+        LockConfig lockConfig = createLockConfig(serverContext.getPluginConfigFilePath());
+        LockClient lockClient =
+                new LockClient(logReplicationNodeId,
+                        lockConfig, runtime.get());
+
+        LockListener lockListener =
+                new LogReplicationLockListener(this);
+
+        LockDataTypes.LockId lockId = LockDataTypes.LockId.newBuilder()
+                .setLockGroup(lockConfig.getLockGroup())
+                .setLockName(lockConfig.getLockName())
+                .build();
+
+        lockClient.getLocks().computeIfAbsent(
+                lockId,
+                key -> new Lock(lockId, lockListener, lockClient.getClientContext(), lockConfig,
+                        initState));
+
+        return lockClient;
+    }
+
+
+    @VisibleForTesting
+    public void deregisterToLogReplicationLock() throws Exception {
+        LockClient lockClient = createLockClient(LockStateType.HAS_LEASE);
+        lockClient.deregisterInterest();
+    }
+
+    @VisibleForTesting
+    public void forceAcquireLogReplicationLock() throws Exception {
```

Comment:
Pavel I'm not sure I understand correctly. How does this force to acquire the lock? let's say this method is called on node A. But node B has the lock, in Corfu Store node B is still marked as the lock owner. How does this enforce? or are you saying that you are faking the state machine to believe it has it. But in practice in the corfu store it does not?
If that is the case how do we control in our tests that two nodes don't have the lock at the same time?

Reply:
I should have documented this somewhere but basically it's just one part for testing. There are multiple issues with testing a leadership change. There are multiple different thread pools in the distributed lock implementation that schedule the periodic tasks all the time, no APIs to control the deregistration/registration to the lock service when needed, fixed timeouts on all the tasks that produce the events for the FSM, and the strong coupling on the discovery service. All of this makes it testing really difficult. I could not find a better way to test the leadership change without rewriting the distributed lock significantly, so I came up with the following, in order to force the leadership change in tests and make them deterministic:

1) Increased these thread pools' timeouts via providing a lock configuration. These will essentially allow deferring the renewals for the lock holder indefinitely as well as postponing the revoking calls from the competitors indefinitely.  
2) Created an API called deregisterToLogReplicationLock - this should be called on the current lock holder. It will force it to release the lock, without renewals (since we have increased these timeouts), and not allow everyone else to get the lock immediately (since we also have increased these timeouts).
3) Created an API called forceAcquireLogReplicationLock - this should be called on some other LR once the previous one has deregistered the interest. 

Doing these three things will guarantee that tests execute exactly the same way all the time without requiring to do a lot of modifications in the actual distributed lock code and breaking something.